src/pybind/mgr/k8sevents/module.py

```diff
@@ -1058,16 +1082,151 @@ def can_run():
+
+            for attr in config_schema[section].keys():
+                if attr not in cfg[section].keys():
+                    raise NameError("Configuration error: Missing {} from {} section".format(attr, section))
+
+                attr_type = config_schema[section][attr]
+                if attr_type == 'https':
+                    # must be a valid https://<wah>:<num> format
+                    res = urlparse(cfg[section][attr])
+                    port = res.netloc.split(":")[-1]
+
+                    if res.scheme != 'https':
+                        raise ValueError("Configuration error: Server URL must use https")
+
+                    elif not res.hostname:
+                        raise ValueError("Configuration error: Invalid server URL format")
+
+                    elif res.hostname:
+                        socket.setdefaulttimeout = 1
+                        try:
+                            socket.gethostbyname(res.hostname)
+                        except socket.gaierror:
+                            raise LookupError("Configuration error: Unresolvable server URL")
+
+                    if not port.isdigit():
+                        raise ValueError("Configuration error: Server URL must end in a port number")
+
+                elif attr_type == 'file':
+                    # must exist
+                    if not os.path.exists(cfg[section][attr]):
+                        raise LookupError("Configuration error: {} file '{}' not found".format(attr, cfg[section][attr]))
+
+                elif attr_type == 'string':
+                    pass
+
+        # Create the API client configuration
+        try:
+            with open(cfg['serviceaccount']['token']) as t:
+                token = t.read().strip('\n')
+        except OSError:
+            raise OSError("Configuration error: Unable to read the token file ({})".format(cfg['serviceaccount']['token']))
+
+        config = client.Configuration()
+        config.host = cfg['connection']['server']
+        config.ssl_ca_cert = cfg['connection']['certificate-authority']
+        config.api_key = { "authorization": "Bearer " + token }
+        api_client = client.ApiClient(config)
+
+        return api_client, cfg['serviceaccount']['namespace']
+
     def serve(self):
+        # apply options set by CLI to this module
         self.config_notify()
 
+        if not kubernetes_imported:
+            self.error_msg = "Unable to start : python kubernetes package is missing"
+        else:
+            if os.environ.get('POD_NAME', None):
+                # running under rook control
+                config.load_incluster_config()
+                self._api = client.CoreV1Api()
+                self._namespace = os.environ.get('POD_NAMESPACE', None)
+            else:
+                # running outside of rook, so we need a config file to tell us how to 
+                # connect to a kubernetes target
+                if os.path.exists(self.kubernetes_config):
```

Comment:
What if the accessing the k8s cluster just works with a proper `$HOME/.kube/config`? Like e.g. in [mgr/rook](https://github.com/ceph/ceph/blob/master/src/pybind/mgr/rook/module.py#L253-L264)

Reply:
I don't think that's not how mgr/rook works. By default, rook uses load_incluster_config. This client function looks for environment variables that point to bindmounted files to pick up the crt and token files . ~/.kube/config is the admin client access mechanism - and IMO we shouldn't be using that for mgr access into K8s resources - since we should be using least privilege principles and pick up the token with access only to the resources we need.

Reply:
let's iron this out on IRC or preferably bluejeans. In the meantime, please review https://github.com/kubernetes-client/python/blob/master/examples/remote_cluster.py as an example of the approach I'm advocating.

Reply:
Looks like we're talking past each other. Sorry for being not clear enough :disappointed: 

> https://github.com/kubernetes-client/python/blob/master/examples/remote_cluster.py as an example of the approach I'm advocating.

Having a dedicated `self.kubernetes_config` is a good thing as it is, as it will simplify the deployment in production environments. So, :+1: for having this!

What if accessing the k8s cluster *already* works with an already working `$HOME/.kube/config`? Would it make sense to simply support `$HOME/.kube/config` in order to ease configuration in simpler environments?

Reply:
@sebastian-philipp latest changes pushed for KV support - please review